### PR TITLE
Fix everything for iOS

### LINKED ISSUE #179: Blur amount is not respected in latest master
Hi there,

I'm testing the latest bits from the master branch and it looks like `blurAmount` property is not respected on iOS. It looks like the issue is from [this change](https://github.com/react-native-community/react-native-blur/commit/71b7e59a627370d3f7d04292777e0ad076a1c869#diff-84ac53a082ccd6040cfdc375b50698cf). 

@eliperkins Do you have an idea what might be wrong?

### PR DESCRIPTION
**Pull request by ndbroadbent**
So, there were a lot of things that needed to be fixed on iOS:

* If you had more than one blur view or vibrancy view (e.g. in the example project), they would all share the same blur amount. There was no way to configure individual blur amounts.
* The vibrancy effect was not updating when the blur effect changed, so the example has actually been broken for a while.
* My changes also fix the problem with `blurAmount` not refreshing during hot reload.
* I fixed the shared `blurAmount` issue with associated objects: http://nshipster.com/associated-objects/
* `VibrancyView` needed exactly the same blur logic from `BlurView`, so it's now a subclass of `BlurView`.

CC: @Kureev @eliperkins 

Closes #179 and #180.

### CODE REVIEW

File: ios/BlurEffectWithAmount.m [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+      [settings setValue:blurAmount forKey:@"blurRadius"];
+    }
+    return settings;
+}
+
+- (id)copyWithZone:(NSZone*)zone
+{
+    id instance = [super copyWithZone:zone];
+    object_setClass(instance, [self class]);
+    // Must also copy blur amount to new instance
```
Comment (eliperkins) [inline ios/BlurEffectWithAmount.m]: Is there a reason we only copy the blur amount and not the blur type or other properties?

File: ios/BlurView.m [hunk 2]
```diff
@@ -1,48 +1,67 @@
...
     if (self = [super initWithFrame:frame]) {
-        self.blurEffect = [UIBlurEffect effectWithStyle:UIBlurEffectStyleDark];
-        self.visualEffectView = [[UIVisualEffectView alloc] initWithEffect:self.blurEffect];
-        self.visualEffectView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
-        self.visualEffectView.frame = frame;
+        self.blurEffectView = [[UIVisualEffectView alloc] init];
+        self.blurEffectView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
+        self.blurEffectView.frame = frame;
+
+        self.blurAmount = [NSNumber numberWithInt:10];
```
Comment (eliperkins) [inline ios/BlurView.m]: What do you think about using NSNumber literals here, like `@10`?

File: ios/BlurView.m [hunk 3]
```diff
@@ -1,48 +1,67 @@
...
+    [self updateBlurEffect];
+  }
+}
+
+
+- (UIBlurEffectStyle)blurEffectStyle
+{
+  if ([self.blurType isEqual: @"xlight"]) return UIBlurEffectStyleExtraLight;
+  if ([self.blurType isEqual: @"light"]) return UIBlurEffectStyleLight;
+  return UIBlurEffectStyleDark;
```
Comment (eliperkins) [inline ios/BlurView.m]: What do you think about having a case here for `@"dark"` as well, just to future proof ourselves? If there were to be another type added in here, we might not find this case so quickly.

File: ios/BlurView.m [hunk 4]
```diff
@@ -1,48 +1,67 @@
...
+}
+
+- (void)updateBlurEffect
+{
+  UIBlurEffectStyle style = [self blurEffectStyle];
+  self.blurEffect = (BlurEffectWithAmount*)[BlurEffectWithAmount
```
Comment (eliperkins) [inline ios/BlurView.m]: Is this cast to `BlurEffectWithAmount` necessary?

File: ios/VibrancyViewManager.m [hunk 5]
```diff
@@ -13,5 +11,6 @@ - (UIView *)view
 }
 
 RCT_EXPORT_VIEW_PROPERTY(blurType, NSString);
+RCT_EXPORT_VIEW_PROPERTY(blurAmount, NSNumber);
```
Comment (eliperkins) [inline ios/VibrancyViewManager.m]: 👍

File: ios/BlurEffectWithAmount.m (hunk 1 above)
Comment (ndbroadbent) [inline ios/BlurEffectWithAmount.m]: Everything else is already copied in `super copyWithZone`, we just need to set the correct class and copy the associated object

File: ios/BlurView.m (hunk 2 above)
Comment (ndbroadbent) [inline ios/BlurView.m]: 👍 (TIL about NSNumber literals!)

File: ios/BlurView.m (hunk 3 above)
Comment (ndbroadbent) [inline ios/BlurView.m]: 👍

File: ios/BlurView.m (hunk 4 above)
Comment (ndbroadbent) [inline ios/BlurView.m]: Oh good point, no that one isn't necessary. It was only necessary for `[BlurEffectWithAmount effectWithStyle:style]`, but not for my custom method.

**Comment by ndbroadbent** [comment]
@eliperkins thanks for the feedback! I've resolved all of those issues.

**Comment by eliperkins** [comment]
Looks good to me! @Kureev, do you mind take a walk through these?

**Pull request merged**